src/main/java/org/aksw/simba/lemming/creation/Inferer.java

```diff
@@ -290,90 +327,93 @@ public OntModel readOntology(String filePath, String base) {
-			T currentResource = stack.pop();
-			String curURI = currentResource.getURI();
-			boolean isSame = false;
-			
-			if (curURI!=null && !uriNodeMap.containsKey(curURI)) {
-				List eqsList = null;
+				//find equivalent classes if possible
+				List<T> eqsList = null;
 				try {
 					if (currentResource.isProperty())
-						eqsList = currentResource.asProperty().listEquivalentProperties().toList();
+						eqsList = (List<T>) currentResource.asProperty().listEquivalentProperties().toList();
 					if (currentResource.isClass())
-						eqsList = currentResource.asClass().listEquivalentClasses().toList();
+						eqsList = (List<T>) currentResource.asClass().listEquivalentClasses().toList();
 				} catch (ConversionException e) {
 					LOGGER.warn(
 							"Cannot convert the equivalents. The ontology does not have any further info on the equivalents of {}.",
 							currentResource.toString());
 				}
 
-				if (eqsList != null && !eqsList.isEmpty()) {
-					stack.addAll(eqsList);
-				}
-				
-				//node to where we want to add the info to
-				Equivalent curNode = null;
-
-				// check to which node do we need to add this info to
-				Iterator<Equivalent> propIterator = elements.iterator();
-				while (propIterator.hasNext()) {
-					curNode = propIterator.next();
-					isSame = curNode.containsElement(currentResource);
-					if (isSame) {
-						curNode.addEquivalent(currentResource);
-						break;
+				Set<T> equiSet = null;
+				if(eqsList == null || eqsList.isEmpty()){
+					if(!iriEquiMap.containsKey(currentResource)){
+						equiSet = new HashSet<>();
+						equiSet.add(currentResource);
+						iriEquiMap.put(currentResource, equiSet);
+					}
+				}else{
+					if(iriEquiMap.containsKey(currentResource)){
+						equiSet = iriEquiMap.get(currentResource);
+					}
+					Map<T, Set<T>> localMap = new HashMap<>();
+					for(T re : eqsList){
+						if(re.getURI()!=null && iriEquiMap.containsKey(re)){
+							localMap.put(re, iriEquiMap.get(re));
+						}
 					}
-				}
 
-				// if not, create new one
-				if (!isSame) {
-					curNode = new Equivalent(currentResource);
-					elements.add(curNode);
-				}
-				
-				// add the node to the map with the URI and add the equivalents (if existing) to the node object
-				if (curNode != null) {
-					uriNodeMap.put(curURI, curNode);
-					if(eqsList != null) {
-						curNode.addEquivalentGroup((Set) eqsList.stream().collect(Collectors.toSet()));
+					for(T re : localMap.keySet()){
+						if(equiSet == null){
+							equiSet = localMap.get(re);
+						}else if (equiSet != localMap.get(re)){
+							equiSet.addAll(localMap.get(re));
+						}
+					}
+
+					if(equiSet==null){
+						equiSet = new HashSet<>();
+						equiSet.add(currentResource);
+					}else{
+						equiSet.add(currentResource);
+					}
+					for(T eq : eqsList){
+						if(eq.getURI() != null){
+							equiSet.add(eq);
+						}
+					}
+					for(T s : equiSet){
+						iriEquiMap.put(s, equiSet);
 					}
 				}

```

Comment:
1) It looks like you don't need the `localMap` at all, since you're just adding the same information as in `iriEquiMap` (L369 `localMap.put(re, iriEquiMap.get(re));`)? 
2) You have 3 loops going through `eqsList` in succession, are these really required?
3) Since all equivalent properties will point to the same `Set<T>`, I think you would only need to update one of the references to this Set, there wouldn't be a need to update every map entry.

Please tell me if I misunderstood something :smiley:

Reply:
The localMap is for the transitive case: if A <=>B <=> C and A, C are visited firstly, then we would get two equiSet (A, B) and (C, B). The function of localMap is to collect such equiSets, and merge them to an object, and update map entries A, B, C using the merged equiSet.

Reply:
Hm from what I see `eqsList` is essentially your set of keys for this `localMap`. 
Then why do you need to have `localMap` if you just add the exact same information present in `iriEquiMap` for this eqsList keyset? Why not work directly with `iriEquiMap`?